cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -55,27 +71,115 @@
+    return new Runnable() {
+      @Override
+      public void run() {
+        for (File file : resources) {
+          if (file == null) {
+            continue;
+          }
+
+          try {
+            if (file.isDirectory()) {
+              DirUtils.deleteDirectoryContents(file);
+            } else {
+              file.delete();
+            }
+          } catch (Throwable t) {
+            LOG.warn("Exception when cleaning up resource {}", file, t);
+          }
+        }
+      }
+    };
+  }
+
+  /**
+   * Creates a local temporary directory for this program run.
+   */
+  private File createTempDirectory(Id.Program programId, RunId runId) {
+    File tempDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                            cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    File dir = new File(tempDir, String.format("%s.%s.%s.%s.%s",
+                                               programId.getType().name().toLowerCase(),
+                                               programId.getNamespaceId(), programId.getApplicationId(),
+                                               programId.getId(), runId.getId()));
+    dir.mkdirs();
+    return dir;
+  }
+
+  /**
+   * Return the copy of the {@link ProgramOptions} including locations of plugin artifacts in it.
+   * @param options the {@link ProgramOptions} in which the locations of plugin artifacts needs to be included
+   * @param programId Id of the Program
+   * @param tempDir Temporary Directory to create the plugin artifact snapshot
+   * @param appSpec program's Application Specification
+   * @return the copy of the program options with locations of plugin artifacts included in them
+   */
+  private ProgramOptions createPluginSnapshot(ProgramOptions options, Id.Program programId, File tempDir,
+                                              @Nullable ApplicationSpecification appSpec) {
+    // appSpec is null in an unit test
+    if (appSpec == null) {
+      return options;
+    }
+
+    Set<String> files = Sets.newHashSet();
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    builder.putAll(options.getArguments().asMap());
+    for (Map.Entry<String, Plugin> pluginEntry : appSpec.getPlugins().entrySet()) {
+      Plugin plugin = pluginEntry.getValue();
+      File destFile = new File(tempDir, ArtifactFile.getFileName(plugin.getArtifactId()));
+      // Skip if the file has already been copied.
+      if (!files.add(destFile.getName())) {
+        continue;
+      }
+
+      try {
+        ArtifactDetail detail = artifactRepository.getArtifact(Id.Artifact.from(programId.getNamespace(),
+                                                                                plugin.getArtifactId()));
+        Files.copy(Locations.newInputSupplier(detail.getDescriptor().getLocation()), destFile);
+      } catch (IOException e) {
+        throw Throwables.propagate(e);
```

Comment:
Better have this method declare `throws IOException` so that the caller need to catch and handle it (e.g. need to run the cleanup task)